nullaway/src/test/java/com/uber/nullaway/NullAwayJSpecifyTests.java

```diff
@@ -200,6 +200,239 @@ public void nullMarkedClassLevelInnerControlsChecking() {
+            "      public void consume(Object s) {",
+            "        System.out.println(s);",
+            "      }",
+            "    };",
+            "  }",
+            "}")
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import com.example.thirdparty.Foo;",
+            "public class Test {",
+            "  public static void test(Object o) {",
+            "    // Safe because IConsumer::consume is unmarked? And no static knowledge of Foo$1",
+            "    Foo.getConsumer().consume(null);",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void nullMarkedOuterMethodLevelWithLocalClass() {
+    defaultCompilationHelper
+        .addSourceLines(
+            "Foo.java",
+            "package com.example.thirdparty;",
+            "import com.uber.nullaway.testdata.annotations.jspecify.future.NullMarked;",
+            "public class Foo {",
+            "  @NullMarked",
+            "  public static String foo(String s) {",
+            "    return s;",
+            "  }",
+            "}")
+        .addSourceLines(
+            "Bar.java",
+            "package com.example.thirdparty;",
+            "import com.uber.nullaway.testdata.annotations.jspecify.future.NullMarked;",
+            "public class Bar {",
+            "  @NullMarked",
+            "  public static Object bar() {",
+            "    class Baz {",
+            "      public void baz() {",
+            "        // BUG: Diagnostic contains: passing @Nullable parameter",
+            "        Foo.foo(null);",
+            "      }",
+            "    }",
+            "    Baz b = new Baz();",
+            "    b.baz();",
+            "    return b;",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void nullMarkedOuterMethodLevelWithLocalClassInit() {
+    defaultCompilationHelper
+        .addSourceLines(
+            "Test.java",
+            "package com.example.thirdparty;",
+            "import com.uber.nullaway.testdata.annotations.jspecify.future.NullMarked;",
+            "public class Test {",
+            "  @NullMarked",
+            "  public static Object test() {",
+            "    class Foo {",
+            "      private Object o;",
+            "      // BUG: Diagnostic contains: initializer method does not guarantee @NonNull field o",
+            "      public Foo() { }",
+            "      public String foo(String s) {",
+            "        return s;",
+            "      }",
+            "    }",
+            "    return new Foo();",
+            "  }",
+            // This currently causes a crash, because updateEnvironmentMapping() is only called when
+            // analyzing the class (which we don't do here, since Test$Foo is unmarked), but used
+            // within analysis of the null-marked method Test$Foo.foo(...).
```

Comment:
This crash is really annoying, since we expect [here](https://github.com/uber/NullAway/blob/master/nullaway/src/main/java/com/uber/nullaway/dataflow/CoreNullnessStoreInitializer.java#L64) (when analyzing methods) information that is only computed [here](https://github.com/uber/NullAway/blob/master/nullaway/src/main/java/com/uber/nullaway/NullAway.java#L1307) (when first visiting the class, if it is annotated). Other than checking all methods eagerly for annotations, I don't know how we know at that point whether to compute that store or not ahead of time, which might mean threading stuff around (particularly the right `TreePath`s) to set up the store lazily.

Reply:
Resolved this by detecting in `NullAway.checkForMethodNullMarkedness` whenever we enter a null-marked method from a null-unmarked context and duplicating the small bit of the logic from `matchClass` that calls `updateEnvironmentMapping(...)` for anonymous/local classes. Not the cleanest arch, but I can't think of an easier solution without refactoring a lot of our dataflow analysis.